internal/gitutil/gitutil.go

```diff
@@ -32,144 +36,282 @@ import (
+// HEAD symref. This is the default branch of the repository.
+func (gur *GitUpstreamRepo) GetDefaultBranch(ctx context.Context) (string, error) {
+	const op errors.Op = "gitutil.GetDefaultBranch"
+	cacheRepo, err := gur.cacheRepo(ctx, gur.URI, []string{}, []string{})
 	if err != nil {
-		return errors.E(op, errors.Git,
-			fmt.Errorf("no 'git' program on path: %w", err))
+		return "", errors.E(op, errors.Repo(gur.URI), err)
 	}
 
-	cmd := exec.Command(p, args...)
-	cmd.Dir = g.Dir
-	cmd.Env = os.Environ()
+	gitRunner, err := NewLocalGitRunner(cacheRepo)
+	if err != nil {
+		return "", errors.E(op, errors.Repo(gur.URI), err)
+	}
 
-	g.Stdout = &bytes.Buffer{}
-	g.Stderr = &bytes.Buffer{}
-	if g.Verbose {
-		// print the command
-		fmt.Println(cmd.Args)
-		cmd.Stdout = io.MultiWriter(g.Stdout, os.Stdout)
-		cmd.Stdout = io.MultiWriter(g.Stderr, os.Stderr)
-	} else {
-		cmd.Stdout = g.Stdout
-		cmd.Stderr = g.Stderr
+	rr, err := gitRunner.Run(ctx, "ls-remote", "--symref", "origin", "HEAD")
+	if err != nil {
+		return "", errors.E(op, errors.Repo(gur.URI), err)
+	}
+	if rr.Stdout == "" {
+		return "", errors.E(op, errors.Repo(gur.URI),
+			fmt.Errorf("unable to detect default branch in repo"))
 	}
 
-	if g.Stdin != nil {
-		cmd.Stdin = g.Stdin
+	re := regexp.MustCompile(`ref: refs/heads/([^\s/]+)\s*HEAD`)
+	match := re.FindStringSubmatch(rr.Stdout)
+	if len(match) != 2 {
+		return "", errors.E(op, errors.Repo(gur.URI), errors.Git,
+			fmt.Errorf("unexpected response from git when determining default branch: %s", rr.Stdout))
 	}
-	err = cmd.Run()
-	if err != nil {
-		return errors.E(op, errors.Git, err)
+	return match[1], nil
+}
+
+// ResolveBranch resolves the branch to a commit SHA. This happens based on the
+// cached information about refs in the upstream repo. If the branch doesn't exist
+// in the upstream repo, the last return value will be false.
+func (gur *GitUpstreamRepo) ResolveBranch(branch string) (string, bool) {
+	if strings.HasPrefix(branch, "refs/heads/") {
+		branch = strings.TrimPrefix(branch, "refs/heads/")
 	}
-	return nil
+	for head, commit := range gur.Heads {
+		if head == branch {
+			return commit, true
+		}
+	}
+	return "", false
+}
+
+// ResolveTag resolves the tag to a commit SHA. This happens based on the
+// cached information about refs in the upstream repo. If the tag doesn't exist
+// in the upstream repo, the last return value will be false.
+func (gur *GitUpstreamRepo) ResolveTag(tag string) (string, bool) {
+	if strings.HasPrefix(tag, "refs/tags/") {
+		tag = strings.TrimPrefix(tag, "refs/tags/")
+	}
+	for t, commit := range gur.Tags {
+		if t == tag {
+			return commit, true
+		}
+	}
+	return "", false
+}
+
+// ResolveRef resolves the ref (either tag or branch) to a commit SHA. If the
+// ref doesn't exist in the upstream repo, the last return value will be false.
+func (gur *GitUpstreamRepo) ResolveRef(ref string) (string, bool) {
+	commit, found := gur.ResolveBranch(ref)
```

Comment:
Any reason for this precedence ? If ref and branch has same name we pull the branch ?